src/gui/SampleThumbnail.cpp

```diff
@@ -0,0 +1,216 @@
+
+/* DEPRECATED; functionality is kept for testing conveniences */
+bool SampleThumbnail::selectFromGlobalThumbnailMap(const Sample& inputSample)
+{
+	const auto samplePtr = inputSample.buffer();
+	const auto name = inputSample.sampleFile();
+	const auto end = s_sampleThumbnailCacheMap.end();
+
+	if (const auto list = s_sampleThumbnailCacheMap.find(name); list != end)
+	{
+		m_thumbnailCache = list->second;
+		return true;
+	}
+
+	m_thumbnailCache = std::make_shared<ThumbnailCache>();
+	s_sampleThumbnailCacheMap.insert(std::make_pair(name, m_thumbnailCache));
+	return false;
+}
+
+/* DEPRECATED; functionality is kept for testing conveniences */
+void SampleThumbnail::cleanUpGlobalThumbnailMap()
+{
+	auto map = s_sampleThumbnailCacheMap.begin();
+	while (map != s_sampleThumbnailCacheMap.end())
+	{
+		// All sample thumbnails are destroyed, a.k.a sample goes out of use
+		if (map->second.use_count() == 1)
+		{
+			s_sampleThumbnailCacheMap.erase(map);
+			map = s_sampleThumbnailCacheMap.begin();
+			continue;
+		}
+
+		map++;
+	}
+}
+
+SampleThumbnail::Thumbnail SampleThumbnail::generate(const size_t thumbnailSize, const SampleFrame* buffer, const size_t size)
+{
+	const auto sampleChunk = (size + thumbnailSize) / thumbnailSize;
+	auto thumbnail = Thumbnail(thumbnailSize);
+
+	for (auto tIndex = size_t{0}; tIndex < thumbnailSize; tIndex++)
+	{
+		const auto
+		sampleIndex = tIndex * size / thumbnailSize,
+		sampleChunkBound = std::min(sampleIndex + sampleChunk, size);
+
+		thumbnail[tIndex] = std::accumulate(buffer + sampleIndex, buffer + sampleChunkBound, Bit{});
+	}
+
+	return thumbnail;
+}
+
+SampleThumbnail::SampleThumbnail(const Sample& inputSample)
+{
+	if (selectFromGlobalThumbnailMap(inputSample)) { return; }
+
+	cleanUpGlobalThumbnailMap();
+
+	const auto sampleBufferSize 	= inputSample.sampleSize();
+	const auto thumbnailSizeDivisor = size_t{10};
+	const auto firstThumbnailSize 	= std::max<size_t>(sampleBufferSize / thumbnailSizeDivisor, 1);
+
+	const auto firstThumbnail 		= generate(firstThumbnailSize, inputSample.data(), sampleBufferSize);
+
+	m_thumbnailCache->push_back(firstThumbnail);
+
+	// Generate the remaining thumbnails using the first one, each one's
+	// size is the size of the previous one divided by the thumbnail size divisor.
+	auto thumbnailSize = size_t{firstThumbnailSize / thumbnailSizeDivisor};
+	while(thumbnailSize >= 1)
+	{
+		const auto& biggerThumbnail = m_thumbnailCache->back();
+
+		auto thumbnail = Thumbnail(thumbnailSize);
+
+		for (auto
+			smallIndex = size_t{0};
+			smallIndex < thumbnail.size();
+			smallIndex += 1
+		) {
+			auto bigIndex 	= smallIndex * thumbnailSizeDivisor;
+			auto bigItStart = biggerThumbnail.cbegin() + bigIndex;
+			auto bigItEnd 	= std::min(bigItStart + thumbnailSizeDivisor, biggerThumbnail.end());
+
+			thumbnail[smallIndex] = std::accumulate(bigItStart, bigItEnd, Bit{});
+		}
+
+		m_thumbnailCache->push_back(thumbnail);
+		thumbnailSize /= thumbnailSizeDivisor;
+	}
```

Comment:
I see that you are calling `generate` only for the first thumbnail, and then doing generation outside of `generate`. This feels a bit awkward. Making `Thumbnail` a class that wraps around the generation code would be nice since all you would need to do is to create the thumbnail and add it to the thumbnail cache.

Reply:
How should we deal with accessing bits from the thumbnail? Do we overload some operators or simply allow access to the the vector that holds the bits?

Reply:
I believe you can overload `operator[]`.

Reply:
Sounds good then